cvat-ui/src/utils/clamp-opacity.ts

```diff
@@ -0,0 +1,26 @@
+// Copyright (C) 2023 CVAT.ai Corporation
+//
+// SPDX-License-Identifier: MIT
+
+import { Job } from 'cvat-core-wrapper';
+import { ShapeType, DimensionType } from 'cvat-core/src/enums';
+import ObjectState from 'cvat-core/src/object-state';
+import { SettingsState } from 'reducers';
+
+export function clampOpacity(
+    states: ObjectState[],
+    shapes: SettingsState['shapes'],
+    job?: Job,
+): [number, number] {
+    const ENHANCED_DEFAULT_OPACITY = 30;
+    const ENHANCED_DEFAULT_SELECTED_OPACITY = 60;
+
+    const withMasks = states
+        .some((_state: ObjectState): boolean => _state.shapeType === ShapeType.MASK);
+    const opacity = withMasks || job?.dimension === DimensionType.DIMENSION_3D ?
+        Math.max(shapes.opacity, ENHANCED_DEFAULT_OPACITY) : shapes.opacity;
+    const selectedOpacity = withMasks || job?.dimension === DimensionType.DIMENSION_3D ?
+        Math.max(shapes.selectedOpacity, ENHANCED_DEFAULT_SELECTED_OPACITY) : shapes.selectedOpacity;
```

Comment:
As an optimization, I would suggest first to check if current opacity/selected opacity already more or equal than enhanced level, just skip the algorithm. Also checking job?.dimension is quicker, so, it is a thing we should probably do first